dali/test/python/test_operator_coord_transform.py

```diff
@@ -0,0 +1,138 @@
+            T = 0
+        if MT is None:
+            MT = 0
+
+        M, T, MT = (x if isinstance(x, dali.data_node.DataNode) else dali.types.Constant(x, dtype=dali.types.FLOAT) for x in (M, T, MT))
+
+        pipe.set_outputs(X, Y, M, T, MT)
+
+    pipe.build()
+    for iter in range(3):
+        outputs = pipe.run()
+        outputs = [x.as_cpu() if hasattr(x, "as_cpu") else x for x in outputs]
+        ref = []
+        scale = 1
+        for idx in range(batch_size):
+            X = outputs[0].at(idx)
+            if T_kind == "fused":
+                MT = outputs[4].at(idx)
+                if MT.size == 1:
+                    M = MT
+                    T = 0
+                else:
+                    M = MT[:,:-1]
+                    T = MT[:,-1]
+            else:
+                M = outputs[2].at(idx)
+                T = outputs[3].at(idx)
+
+            if M.size == 1:
+               Y = X.astype(np.float32) * M + T
+            else:
+               Y = np.matmul(X.astype(np.float32), M.transpose()) + T
+
+            if np.issubdtype(out_dtype, np.integer):
+                info = np.iinfo(out_dtype)
+                Y = Y.clip(info.min, info.max)
+
+            ref.append(Y)
+            scale = max(scale, np.max(np.abs(Y)) - np.min(np.abs(Y)))
+        avg = 1e-6 * scale
+        eps = 1e-6 * scale
+        if out_dtype != np.float32:  # headroom for rounding
+            avg += 0.33
+            eps += 0.5
+        check_batch(outputs[1], ref, batch_size, eps, eps, compare_layouts=False)
+
+
+def test_all():
+    for device in ["cpu", "gpu"]:
+        for M_kind in [None, "vector", "scalar", "input", "scalar input"]:
+            for T_kind in [None, "vector", "scalar", "input", "scalar input"]:
+                for batch_size in [1,3]:
+                    yield _run_test, device, batch_size, 3, 3, np.float32, np.float32, M_kind, T_kind
+
+    for device in ["cpu", "gpu"]:
+        for in_dtype in [np.uint8, np.uint16, np.int16, np.int32, np.float32]:
+            for out_dtype in set([in_dtype, np.float32]):
+                for batch_size in [1,8]:
+                    yield _run_test, device, batch_size, 3, 3, in_dtype, out_dtype, "input", "input"
+
+    for device in ["cpu", "gpu"]:
+        for M_kind in ["input", "vector", None]:
+            for in_dim in [1,2,3,4,5,6]:
+                out_dims = [1,2,3,4,5,6] if M_kind == "vector" or M_kind == "input" else [in_dim]
+                for out_dim in out_dims:
+                    yield _run_test, device, 2, out_dim, in_dim, np.float32, np.float32, M_kind, "vector"
+
+    for device in ["cpu", "gpu"]:
+        for MT_kind in ["vector", "input", "scalar"]:
+            for in_dim in [1,2,3,4,5,6]:
+                out_dims = [1,2,3,4,5,6] if MT_kind == "vector" or MT_kind == "input" else [in_dim]
+                for out_dim in out_dims:
+                    yield _run_test, device, 2, out_dim, in_dim, np.float32, np.float32, MT_kind, "fused"
+
```

Comment:
Can you add a set of negative cases for invalid M, T and MT values?

Reply:
I'm too close to the dreaded 1000 lines... ;)

Reply:
that shouldn't be an excuse to avoid adding tests :)

Reply:
that shouldn't be an excuse to avoid adding tests :)